source/simplex/quadrature_lib.cc

```diff
@@ -263,6 +159,248 @@ namespace Simplex
              Utilities::to_string(n_points_1D)));
   }
 
+  namespace
+  {
+    template <std::size_t b_dim>
+    std::vector<std::array<double, b_dim>>
+    all_permutations(const std::array<double, b_dim> &b_point)
+    {
+      std::vector<std::array<double, b_dim>> output;
+
+      // We want all possible permutations of the barycentric coordinates.
+      // The easiest way to get all of them is to sort the input first and
+      // then use next_permutation to cycle through them all.
+      std::array<double, b_dim> temp = b_point;
+      std::sort(temp.begin(), temp.end());
+      do
+        {
+          output.push_back(temp);
+        }
+      while (std::next_permutation(temp.begin(), temp.end()));
+
+      return output;
+    }
+  } // namespace
+
+
+
+  template <int dim>
+  QWitherdenVincent<dim>::QWitherdenVincent(const unsigned int n_points_1D)
+    : QSimplex<dim>(Quadrature<dim>())
+  {
+    // Just use Gauss in 1D: this is a high-order open rule so this is a
+    // reasonable equivalent for generic programming.
+    if (dim == 1)
+      {
+        const dealii::QGauss<dim> quad(n_points_1D);
+        this->quadrature_points = quad.get_points();
+        this->weights           = quad.get_weights();
+        return;
+      }
+
+    std::array<double, dim + 1> centroid;
+    std::fill(centroid.begin(), centroid.end(), 1.0 / (dim + 1.0));
+    std::vector<std::vector<std::array<double, dim + 1>>> b_point_permutations;
+    std::vector<double>                                   b_weights;
+
+    // We can simplify the implementation of these quadrature rules
+    // by quite a bit by exploiting symmetry - we do essentially the
+    // same thing for each barycentric coordinate, so we can express
+    // our quadrature rule as permutations of barycentric points
+    // instead of writing things out explicitly.
+
+    // Apply a Barycentric permutation where one point is different.
+    auto process_point_1 = [&](const double a, const double w) {
+      const double                b = 1.0 - dim * a;
+      std::array<double, dim + 1> b_point;
+      std::fill(b_point.begin(), b_point.end(), a);
```

Comment:
```suggestion
      std::fill(b_point.begin(), b_point.end()-1, a);
```

Reply:
Or, maybe to make the point more clear:
```suggestion
      std::fill(b_point.begin(), b_point.begin()+dim, a);
```